Add ptest for measuring and reporting the total boot time

gratian opened the pull request:
Add an NI-specific ptest for logging the total boot time and uploading the data to the NI-internal influxdb data base.

### Testing

- [x] bitbake ni-boot-time
- [x] bitbake packagegroup-ni-ptest-smoke
- [x] installed from local feed and manually run on `cRIO-9042`, and `PXIe-8840`
- [x] verified that the data reported is available via Grafana/InfluxDB
- [x] verified reasonable behavior on error: the test fails if the boot time log is missing or the influxdb connection is not available

gratian committed:
ni-test-boot-time: add ptest for publishing boot time data to influxdb

Add a ptest for publishing the NILRT boot time to the NI internal influxdb
instance.

For implementation simplicity reasons, since the ptest framework cannot run
a test across a reboot, the test relies on apriori knowledge about the NI
ATS system - specifically it assumes a reboot is issued between installing
ptest packages and before running the tests via 'ptest-runner'.

This lets us split the test steps into:

  - an init script called 'zz-ni-record-boot-time' configured to run as the
    last script in /etc/rc[4,5].d run levels and save the total boot time
    to a log file under '/var/log/boottime'.

  - a separate reporting script run by 'ptest-runner' which reads the log
    file saved at boot and uploads the boot time data together with device
    specific metadata to the NI influxdb instance.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
packagegroup-ni-ptest-smoke: add ni-test-boot-time ptest

Add ptest for measuring and reporting total boot time.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

amstewart commented on recipes-ni/ni-boot-time/ni-boot-time.bb:
```diff
@@ -0,0 +1,39 @@
+	install -m 0755 ${S}/upload_results.py	${D}${PTEST_PATH}
+
+	install -d ${D}${sysconfdir}/init.d
+	install -m 0755 ${S}/zboottime		${D}${sysconfdir}/init.d
```
Can we give this initscript a more informative name, like `zz-ni-record-boottime`?

amstewart commented on recipes-ni/ni-boot-time/ni-boot-time.bb:
```diff
@@ -0,0 +1,39 @@
+}
+
+pkg_postinst_ontarget:${PN}-ptest:append() {
+	python3 -m pip install influxdb
```
Do we not have an influxdb python3 package in the feeds? Is that intentional, or have we just not gotten around to making one?

gratian replied:
We do not and I'm not finding an upstream recipe in the OE layer index either. There is a recipe  for the infulxdb i.e. the database itself but not the python package that interfaces with it.

It's not intentional, just not gotten around to it; arguably it should be added to the backlog.

amstewart replied:
Could you make a TD item for it? Otherwise, I suppose it isn't blocking for this PR.

gratian replied:
TD #2358572

amstewart commented on recipes-ni/ni-boot-time/ni-boot-time.bb:
```diff
@@ -0,0 +1,39 @@
+SUMMARY = "Ptest for measuring and logging the boot time"
```
At this point, there are several performance tests in the smoke-ptests group that all upload to our influxdb server; right? Can we merge these all into a single `ni-performance-testing` package?

gratian replied:
Actually this is the first performance test that we've added to smoke-ptests. I've added it there because it is quick and cheap to run with the other smoke tests.

The other set of tests that upload to influxdb are the kernel-performance-tests which are run separately (due to how long they take) and are not part of smoke-ptests.

amstewart replied:
Hmm.. OK. Could we rename this recipe to something which makes it more clear that it is a test package then? Maybe `ni-test-boott-time`?

amstewart commented on recipes-ni/ni-boot-time/files/upload_results.py:
```diff
@@ -0,0 +1,137 @@
+import subprocess
+from datetime import datetime
+from influxdb import InfluxDBClient, exceptions
+import requests.exceptions
```
`requests` isn't python-standard I think. Are you missing an RDEPENDS for this?

gratian replied:
I'll add an RDEPENDS. Fortunately we do have `python3-requests` in the feed already.
FWIW it works w/o RDEPENDS because it gets pip installed as a dependency of `influxdb`.

gratian commented:
Changes since v1:

- renamed:    zboottime -> zz-ni-record-boottime
- renamed:    recipes-ni/ni-boot-time/ni-boot-time.bb -> recipes-ni/ni-test-boot-time/ni-test-boot-time.bb and changed all associated references
- added RDEPENDS for `python3-requests`

gratian commented:
v2 testing

- [x] bitbake packagegroup-ni-ptest-smoke
- [x] opkg install packagegroup-ni-ptest-smoke (from local feed)
- [x] ptest-runner ni-test-boot-time

jpautler commented on recipes-ni/ni-test-boot-time/files/zz-ni-record-boot-time:
I think there was originally some discussion about building the write-boot-time-to-logfile behavior directly into our image. That way it would always be present and could be used by anyone for debugging/etc. It could also be the first of potentially other metrics to log (time until network is up, etc). Comments?

gratian replied:
The way things are currently architected you can install the base package (i.e. not the -ptest) and get the log:

```
admin@NI-PXIe-8840-00802f23aed1:~# opkg files ni-test-boot-time
Package ni-test-boot-time (1.0-r0.0) is installed on root and has the following files:
/etc/rc4.d/S99zz-ni-record-boot-time
/etc/init.d/zz-ni-record-boot-time
...
/etc/rc5.d/S99zz-ni-record-boot-time
```

We can decide to add this package to the base image however I would like to do that in a separate PR and have a discussion about it. I personally do not think its a good idea for several reasons:

- it is a distro/linux non-standard log.
- it will slow down the boot for everybody and most customers won't know or care about this log.
- it is not a very useful way to measure boot time because it doesn't tell you anything about what is causing the slow down. For that we have more useful tools in the feed, both `bootchart` and `bootchart2` are available.
- if somebody really wants it for their test pipelines they can `opkg install ni-test-boot-time`.

jpautler commented on recipes-ni/ni-test-boot-time/files/upload_results.py:
```diff
@@ -0,0 +1,137 @@
+
+    data = {
+        "measurement": "boot_performance",
+        "tags": {
```
Would it make sense to log a device hostname/serial number? I'm just imagining what kind of info would be useful if we see an unexpected change. Being able to trace a change to a recently swapped controller-under-test could be helpful.

gratian replied:
I don't expect large variations for the same controller model unless there's a hardware issue but I can see how having this info would make tracking a hardware issue easier so I'll add it.

jpautler commented on recipes-ni/ni-test-boot-time/files/run-ptest:
```diff
@@ -0,0 +1,21 @@
+	RESULT=$?
+else
+	echo "INFO: InfluxDB connection information not found at '$INFLUXDB_INFO'"
+	RESULT=1
```
For local ptest runs or runs of debug dashboard events (events with a "debug" tag), there will be no .influxdb.info file present on the target. I don't know if it's better for this test to "pass" in that case? This test is a little different from some other tests which have a logical pass/fail result independent from whether the data is uploaded or not.

gratian replied:
Since the main action of this test is uploading the boot time to influxdb for later comparisons I have decided it is better to fail the test if the upload fails. That way we would catch if something is wrong and we're not actually capturing results.

If you prefer, I could potentially "skip" the test in the cases where the influxdb info is not present. I am reluctant to always "pass" the test since it is going to hide cases were the reporting stopped working and we won't notice until much later when we see the data is missing.

Long term once we collect enough data from all devices we can consider adding device specific .conf files to this package with thresholds to test against and report a more accurate pass/fail. At this time we don't have enough data to establish these pass/fail thresholds.

gratian replied:
I went ahead and marked the test SKIP in this situation (see v3).

jpautler commented on recipes-ni/ni-test-boot-time/files/upload_results.py:
```diff
@@ -0,0 +1,137 @@
+        if not influxdb.write_points(data):
+            logging.warning("failed to write data points to influxdb")
+        influxdb.close()
+    except (exceptions.InfluxDBClientError,
```
I believe an exception here will not trigger the ptest to fail since you are catching and not re-throwing. I don't know if that's what you intend. Since at this point you will have found influxdb info in /home/admin/.influxdb.info, maybe it should cause a failure?

gratian commented:
Changes since v2:

- mark test as SKIP instead of FAIL if the influxdb connection information is not present
- cleaned-up exception handling and re-tested error paths. The test will now fail if the influxdb connection is present but the upload script cannot connect to the server (see jpautler's comment above about re-throwing the exception)
- added metadata for hostname and serial# to the reported data

v3 testing:

- [x] bitbake packagegroup-ni-ptest-smoke
- [x] opkg install ni-test-boot-time-ptest (from local feed)
- [x] ptest-runner
    - [x] verified test is skipped if `.influxdb.info` is missing
    - [x] verified test fails if various error/exceptions paths are activated
    - [x] verified the test now reports hostname and serial# information

amstewart commented:
I assume this should be cherry-picked to kirkstone as well.

amstewart commented:
Cherry-picked to kirkstone.
